### DISCUSSION TITLE
End to end typesafety?

### DISCUSSION THREAD

**Discussion by trompx** [porsager/postgres#442] [General]
Hello, I was wondering if postgres.js provide end to end typesafety just like prisma, kisely, slonik, zapatos etc. do?
Or if the typescript current implementation were not providing autocompletion on the frontend?

**Comment by Minigugus** [comment]
`postgres` only deals with communication with PostgreSQL, just like `pq`. TS types allows to specify query return types, but have to be passed manually on each query (not inferred from the database).

There are some projects like [`ts-sql-plugin`](https://github.com/xialvjun/ts-sql-plugin) that aim to provide realtime completion and validation based on a live database from the tagged template string syntax, I haven't tested it on `postgres` so I can't tell if it is compatible.
There is also [`postgres-ts-plugin`](https://github.com/Minigugus/postgres-ts-plugin#readme) for `postgres` specifically but I don't have much time to maintain it (help welcomed)
